client/consensus/grandpa/finality_proof.go

```diff
@@ -0,0 +1,248 @@
+// Copyright 2023 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"errors"
+
+	"github.com/ChainSafe/gossamer/pkg/scale"
+	"golang.org/x/exp/constraints"
+)
+
+// GRANDPA block finality proof generation and check.
+//
+// Finality of block B is proved by providing:
+// 1) the justification for the descendant block F;
+// 2) headers sub-chain (B; F] if B != F;
+// 3) proof of GRANDPA::authorities() if the set changes at block F.
+//
+// Since earliest possible justification is returned, the GRANDPA authorities set
+// at the block F is guaranteed to be the same as in the block B (this is because block
+// that enacts new GRANDPA authorities set always comes with justification). It also
+// means that the `set_id` is the same at blocks B and F.
+//
+// Let U be the last finalized block known to caller. If authorities set has changed several
+// times in the (U; F] interval, multiple finality proof fragments are returned (one for each
+// authority set change) and they must be verified in-order.
+//
+// Finality proof provider can choose how to provide finality proof on its own. The incomplete
+// finality proof (that finalises some block C that is ancestor of the B and descendant
+// of the U) could be returned.
+
+var (
+	// The requested block has not yet been finalized
+	errBlockNotYetFinalized = errors.New("block not yet finalized")
+	// The requested block is not covered by authority set changes. Likely this means the block is
+	// in the authoritySetChangeIDLatest authority set, and the subscription API is more appropriate
+	errBlockNotInAuthoritySetChanges = errors.New("block not covered by authority set changes")
+)
+
+const maxUnknownHeaders = 100_000
+
+// FinalityProofProvider Finality proof provider for serving network requests.
+type FinalityProofProvider[
+	BE Backend[Hash, N, H, B],
+	Hash constraints.Ordered,
+	N constraints.Unsigned,
+	AuthID AuthorityID,
+	S comparable,
+	ID AuthorityID,
```

Comment:
I did not need ID (good catch) as it was a duplicate, but I do need the signature because the constraint gets used to retrieve the best justification in `proveFinality`